test/object-store/sync/app.cpp

```diff
@@ -2089,6 +2090,11 @@ TEST_CASE("app: sync integration", "[sync][app]") {
         std::mutex sync_error_mutex;
         util::Optional<SyncError> sync_error;
         config.sync_config->error_handler = [&](auto, SyncError error) {
+            if (error.error_code.category() != util::websocket::websocket_close_status_category()) {
+                // Ignore any errors not related to this test.
+                return;
```

Comment:
I'm not familiar with the other errors occurring here. But won't this change turns the test into something that occasionally hangs instead of something that occasionally fails? It seems like the underlying problem is not being addressed. What would you think about also adding a timeout to the main loop?

Reply:
So, I added a timeout to the main loop, but your comment got me curious about whether there were always more than one error that the test was racing with, so I made the sync_errors a vector and collected all of them and changed the test so that it waited for the size of the vector to be  greater than 1 and contain at least one error of the category this test cares about. On my Mac:
```
(std::vector<realm::SyncError, std::allocator<realm::SyncError> >) $0 = size=6 {
  [0] = {
    error_code = (__val_ = 32, __cat_ = 0x000000010221ebf8)
    message = "Broken pipe"
    is_fatal = false
    user_info = size=0 {}
    is_unrecognized_by_client = true
  }
  [1] = {
    error_code = (__val_ = 32, __cat_ = 0x000000010221ebf8)
    message = "Broken pipe"
    is_fatal = false
    user_info = size=0 {}
    is_unrecognized_by_client = true
  }
  [2] = {
    error_code = (__val_ = 32, __cat_ = 0x000000010221ebf8)
    message = "Broken pipe"
    is_fatal = false
    user_info = size=0 {}
    is_unrecognized_by_client = true
  }
  [3] = {
    error_code = (__val_ = 32, __cat_ = 0x000000010221ebf8)
    message = "Broken pipe"
    is_fatal = false
    user_info = size=0 {}
    is_unrecognized_by_client = true
  }
  [4] = {
    error_code = (__val_ = 32, __cat_ = 0x000000010221ebf8)
    message = "Broken pipe"
    is_fatal = false
    user_info = size=0 {}
    is_unrecognized_by_client = true
  }
  [5] = {
    error_code = (__val_ = 1009, __cat_ = 0x000000010221ec30)
    message = "read limited at 16777217 bytes"
    is_fatal = false
    user_info = size=0 {}
    is_unrecognized_by_client = true
  }
```
On the ASAN builder that originally uncovered the flakeyness:
```
(gdb) p sync_errors
$1 = std::vector of length 4, capacity 4 = {{error_code = {_M_value = 104,
      _M_cat = 0x34d20a0 <(anonymous namespace)::g_system_category>}, message = <incomplete type>, is_fatal = false,
    user_info = std::unordered_map with 0 elements, is_unrecognized_by_client = true, static c_original_file_path_key =
    "ORIGINAL_FILE_PATH", static c_recovery_file_path_key = <same as static member of an already seen type>}, {error_code = {
      _M_value = 104, _M_cat = 0x34d20a0 <(anonymous namespace)::g_system_category>}, message = <incomplete type>, is_fatal = false,
    user_info = std::unordered_map with 0 elements, is_unrecognized_by_client = true,
    static c_original_file_path_key = "ORIGINAL_FILE_PATH",
    static c_recovery_file_path_key = <same as static member of an already seen type>}, {error_code = {_M_value = 104,
      _M_cat = 0x34d20a0 <(anonymous namespace)::g_system_category>}, message = <incomplete type>, is_fatal = false,
    user_info = std::unordered_map with 0 elements, is_unrecognized_by_client = true,
    static c_original_file_path_key = "ORIGINAL_FILE_PATH",
    static c_recovery_file_path_key = <same as static member of an already seen type>}, {error_code = {_M_value = 1009,
      _M_cat = 0x34dd120 <realm::util::websocket::websocket_close_status_category()::category>}, message = <incomplete type>,
    is_fatal = false, user_info = std::unordered_map with 0 elements, is_unrecognized_by_client = true,
    static c_original_file_path_key = "ORIGINAL_FILE_PATH",
    static c_recovery_file_path_key = <same as static member of an already seen type>}}
(gdb)
```

I think these broken pipe errors and ECONNRESET errors are benign from the test's perspective. Either way, I also added a timeout so this won't hang for forever.

Reply:
I don't know how to interpret the pipe errors in this situation without doing a deep investigation, sorry.